crates/bevy_utils/src/staging.rs

```diff
@@ -0,0 +1,1108 @@
+    #[inline]
+    fn stage_scope_locked<R>(
+        &mut self,
+        f: impl FnOnce(&mut Stager<<Self::Core as StagableWritesTypes>::Staging>) -> R,
+    ) -> R {
+        f(&mut self.stage_lock().as_stager())
+    }
+}
+
+impl<T: StagableWritesCore> StagableWrites for T {
+    type Core = T;
+
+    fn get_core(&self) -> &Self::Core {
+        self
+    }
+}
+
+/// A struct that allows staging changes while reading from cold storage.
+/// Generally, staging changes should be implemented on this type.
+#[derive(Debug)]
+pub struct Stager<'a, T: StagedChanges> {
+    /// The storage that is read optimized.
+    pub cold: &'a T::Cold,
+    /// The staged changes.
+    pub staged: &'a mut T,
+}
+
+/// A struct that allows accessing changes while reading from cold storage.
+/// Generally, reading data should be implemented on this type.
+#[derive(Debug)]
+pub struct StagedRef<'a, T: StagedChanges> {
+    /// The storage that is read optimized.
+    pub cold: &'a T::Cold,
+    /// The staged changes.
+    pub staged: &'a T,
+}
+
+/// A locked version of [`Stager`].
+/// Use this to hold a lock guard while using [`StagerLocked::as_stager`] or similar.
+#[derive(Debug)]
+pub struct StagerLocked<'a, T: StagableWritesCore> {
+    inner: RefStageOnWrite<'a, T>,
+    /// The storage that is read optimized.
+    pub cold: T::ColdRef<'a>,
+    /// The staged changes.
+    pub staged: RwLockWriteGuard<'a, T::Staging>,
+}
+
+/// A locked version of [`StagedRef`].
+/// Use this to hold a lock guard while using [`StagerLocked::as_staged_ref`].
+#[derive(Debug)]
+pub struct StagedRefLocked<'a, T: StagableWritesCore> {
+    inner: RefStageOnWrite<'a, T>,
+    /// The storage that is read optimized.
+    pub cold: T::ColdRef<'a>,
+    /// The staged changes.
+    pub staged: RwLockReadGuard<'a, T::Staging>,
+}
+
+/// A general purpose enum for representing data that may or may not need to be staged.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum MaybeStaged<C, S> {
+    /// There is no staging necessary.
+    Cold(C),
+    /// There is staging necessary.
+    Staged(S),
+}
+
+/// A struct that allows read-optimized operations while still allowing mutation.
+/// When mutations are made, they are staged.
+/// Then, at user-defined times, they are applied to the read-optimized storage.
+/// This allows mutations through [`RwLock`]s without needing to constantly lock old or cold data.
+///
+/// This is not designed for atomic use (ie. in an `Arc`).
+#[cfg_attr(feature = "alloc", doc = "See [`AtomicStageOnWrite`] for that.")]
+#[derive(Default, Debug)]
+pub struct StageOnWrite<T: StagedChanges> {
+    /// Cold data is read optimized.
+    cold: T::Cold,
+    /// Staged data stores recent modifications to cold. It's [`RwLock`] coordinates mutations.
+    staged: RwLock<T>,
+}
+
+/// A version of [`StageOnWrite`] designed for atomic use.
+/// It functions fully without needing `&mut self`.
+/// See [`StageOnWrite`] for details.
```

Comment:
I think this line is redundant. We already link to it on the first line.

Repeat for ArcStageOnWrite.